<script type="text/javascript">
    (function($){
        $( document ).ready(function() {
            var webpart = <%=webpartContext%>;

            var childEl = jQuery('#' + webpart.wrapperDivId + ' > div.wp');
            childEl.attr('id', webpart.wrapperDivId + '-wp');

            LDK.Utils.getReadOnlyQWP({
                frame: 'portal',
                title: 'List of mGAP Functional Annotations',
                name: 'query',
                schemaName: 'mGAP',
                queryName: 'annotations',
                maxRows: 20
            }).render(childEl.attr('id'));
        });
    })(jQuery);
</script>

<div style="float: none;">
    <div style="float: left; border: 10px; padding-right: 20px;width:40%; text-align: justify" >
        Assigning significance to genetics variants is often a challenge, and this is particularly problematic for model organisms like macaques,
        which have few sources of genomic annotations, relative to human or mouse.
        To overcome this, we developed a multipart process to annotate our data. <br><br>NHP data are aligned to the latest NCBI genomic build for the species.
        We use liftOver tools to lift variants to the human genome (GRCh37), with ~65% of variants typically being successfully lifted. These variants are annotated using a variety of human data sources, described below.
        These annotated variants are translated back to genomic coordinates of the NHP species, and all data sources merged. In addition to human annotations, SnpEff is used to annotate predicted impact on protein coding (run using the Ensembl gene build).
        <br>
        <br>
        Our annotation sources are drawn from a number of databases and tools. A complete field-by-field list is shown in the table at the bottom of the page. Please note that in most cases the 'Data Source' column is a link to the primary data, which may have more information on that annotation.
        <p />
        <ul>
            <li><a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035889931-Funcotator-Information-and-Tutorial">Funcotator (which is a tool for annotation and an aggregation of data sources)</a></li>
            <li><a href="http://database.liulab.science/dbNSFP#intro">dbNSFP: Functional Predictions and Annotations for Human Non-synonymous and Splice Site SNVs</a></li>
            <li><a href="https://pcingola.github.io/SnpEff/se_introduction/">SnpEff: variant annotation and effect prediction tool</a></li>
            <li><a href="https://github.com/bimberlabinternal/VariantAnnotation/tree/master">VariantAnnotation: a custom Funcotator-compatible set of annotation sources</a></li>

        </ul>
    </div>
    <div>
        <img width="600 px;" src="<%=contextPath%>/mgap/images/annotation.png" alt="Annotation Strategy">
    </div>
</div>

<br>
<br>
<br>
<div class="wp" ></div>
